server/terminal/terminal.go

```diff
@@ -0,0 +1,236 @@
+		return
+	}
+
+	ctx := r.Context()
+	a, err := s.appLister.Get(app)
+	if err != nil {
+		http.Error(w, "Cannot get app", http.StatusBadRequest)
+		return
+	}
+
+	if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplications, rbacpolicy.ActionGet, appRBACName(*a)); err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	if err := s.enf.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceExec, rbacpolicy.ActionGet, appRBACName(*a)); err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	config, err := s.getApplicationClusterRawConfig(ctx, a)
+	if err != nil {
+		http.Error(w, "Cannot get raw cluster config", http.StatusBadRequest)
+		return
+	}
+
+	kubeClientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		http.Error(w, "Cannot initialize kubeclient", http.StatusBadRequest)
+		return
+	}
+
+	resourceTree, err := s.appResourceTreeFn(ctx, a)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// From the tree find pods which match the given pod.
+	if !podExists(resourceTree.Nodes, podName, namespace) {
+		http.Error(w, "Pod doesn't belong to specified app", http.StatusBadRequest)
+		return
+	}
+
+	pod, err := kubeClientset.CoreV1().Pods(namespace).Get(ctx, podName, metav1.GetOptions{})
+	if err != nil {
+		http.Error(w, "Cannot find pod", http.StatusBadRequest)
+		return
+	}
+
+	if pod.Status.Phase != v1.PodRunning {
+		http.Error(w, "Pod not running", http.StatusBadRequest)
+		return
+	}
+
+	var findContainer bool
+	for _, c := range pod.Spec.Containers {
+		if container == c.Name {
+			findContainer = true
+			break
+		}
+	}
+	if !findContainer {
+		http.Error(w, "Cannot find container", http.StatusBadRequest)
+		return
+	}
+
+	session, err := newTerminalSession(w, r, nil)
+	if err != nil {
+		http.Error(w, "Failed to start terminal session", http.StatusBadRequest)
+		return
+	}
+	defer session.Done()
+
+	validShells := []string{"bash", "sh", "powershell", "cmd"}
+	if isValidShell(validShells, shell) {
+		cmd := []string{shell}
+		err = startProcess(kubeClientset, config, namespace, podName, container, cmd, session)
+	} else {
+		// No shell given or it was not valid: try some shells until one succeeds or all fail
+		// FIXME: if the first shell fails then the first keyboard event is lost
+		for _, testShell := range validShells {
+			cmd := []string{testShell}
+			if err = startProcess(kubeClientset, config, namespace, podName, container, cmd, session); err == nil {
+				break
+			}
+		}
+	}
+
+	if err != nil {
+		http.Error(w, "Failed to exec container", http.StatusBadRequest)
+		session.Close()
+		return
+	}
+
+	session.Close()
+}
+
+// appRBACName formats fully qualified application name for RBAC check
+func appRBACName(app appv1.Application) string {
+	return fmt.Sprintf("%s/%s", app.Spec.GetProject(), app.Name)
+}
+
+func podExists(treeNodes []appv1.ResourceNode, podName, namespace string) bool {
```

Comment:
Can we add unit tests for this function? I'm mostly interested in testing that empty UIDs return false, since I think resources can appear in the resource tree without a UID even if they shouldn't be managed by the app. @alexmt can tell me whether I'm remembering that correctly.

Reply:
I added the unit tests.